File: src/python/library/yocrypt/decode.py
import string


def decode(text, **kwargs):
    text = text
    key = text.split()[0]
    for identifier, value in kwargs.items():
        if identifier == "key":
            key = value
            text = value + " " + text

    result = list(text)
    key_text = []

    while len(key_text) < len(text):
        key_text.append(key)

    key_text = list("".join(key_text))

    for i, letter in enumerate(text):
        if letter in string.ascii_lowercase:
            continue
        key_text.insert(i, letter)

    key_text = key_text[:len(text)]

    for i, letter in enumerate(key_text):
        if not letter in string.ascii_lowercase:
            continue
        if not result[i] in string.ascii_lowercase:
            continue
        a = string.ascii_lowercase.index(
            result[i])-string.ascii_lowercase.index(letter)-1
        while a >= 26:
            a -= 26
        result[i] = string.ascii_lowercase[a]

    return key + " " + "".join(result[len(key)+1:])

File: src/javascript/library/yocrypt/decode.js
module.exports = function (text, key = null) {
    const letters = "abcdefghijklmnopqrstuvwxyz";
    if (key == null) {
        key = text.split(' ')[0];
    }
    let result = text.split('');
    let key_text = [];

    while (key_text.length < text.length) {
        key_text.push(key);
    }

    key_text = key_text.join("").split('');
    for (const [i, letter] of result.entries()) {
        if (letters.includes(letter)) {
            continue;
        }
        key_text.splice(i, 0, letter);
    }

    key_text = key_text.slice(0, text.length);

    for (const [i, letter] of key_text.entries()) {
        if (!letters.includes(letter)) {
            continue;
        }
        if (!letters.includes(result[i])) {
            continue;
        }
        let a = letters.indexOf(result[i]) - letters.indexOf(letter) - 1;
        while (a >= 26) {
            a -= 26;
        }
        if (a < 0) {
            a = 26 + a;
        }
        result[i] = letters[a];
    }

    return key + " " + result.slice(key.length + 1).join("");
}